packages/web3-test/src/token.ts

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2018 - 2022 The Alephium Authors
+This file is part of the alephium project.
+
+The library is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+The library is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with the library. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+import { Contract, DUST_AMOUNT, NodeProvider, ONE_ALPH, Script, SignerProvider, stringToHex } from '@alephium/web3'
+
+function createTokenContract(symbol: string, name: string): string {
+  return `
+    Contract Token(totalSupply: U256) implements IFungibleToken {
+      pub fn getSymbol() -> ByteVec {
+        return #${stringToHex(symbol)}
+      }
+
+      pub fn getName() -> ByteVec {
+        return #${stringToHex(name)}
+      }
+
+      pub fn getDecimals() -> U256 {
+        return 18
+      }
+
+      pub fn getTotalSupply() -> U256 {
+        return totalSupply
+      }
+    }
+
+    @std(id = #0001)
+    Interface IFungibleToken {
+      pub fn getSymbol() -> ByteVec
+
+      pub fn getName() -> ByteVec
+
+      pub fn getDecimals() -> U256
+
+      pub fn getTotalSupply() -> U256
+    }
+  `
+}
+
+const contractCode = createTokenContract('TestToken', 'TestToken')
+
+let _contract: Contract | undefined = undefined
+
+async function getContractArtifact(nodeProvider: NodeProvider): Promise<Contract> {
+  if (_contract !== undefined) return _contract
+  const compileResult = await nodeProvider.contracts.postContractsCompileContract({ code: contractCode })
+  _contract = Contract.fromCompileResult(compileResult)
+  return _contract
+}
+
+async function getScriptArtifact(nodeProvider: NodeProvider): Promise<Script> {
+  const contract = await getContractArtifact(nodeProvider)
+  const scriptCode = `
+    TxScript Main(address: Address, totalSupply: U256) {
+      let (encodedImmFields, encodedMutFields) = Token.encodeFields!(totalSupply)
+      createContractWithToken!{address -> ALPH: 1 alph}(
+        #${contract.bytecode},
+        encodedImmFields,
+        encodedMutFields,
+        totalSupply,
+        address
+      )
+    }
+    ${contractCode}
+  `
+  const scriptResult = await nodeProvider.contracts.postContractsCompileScript({ code: scriptCode })
+  return Script.fromCompileResult(scriptResult)
+}
+
+export async function createAndTransferToken(nodeProvider: NodeProvider, signer: SignerProvider, amount: bigint) {
```

Comment:
1. We should use the default addresses to deploy the script and execute the script, so that the balance of recipient is not changed. Otherwise it's hard to test the balance changes
2. signer should be renamed as `recipient: Address`

Reply:
Currently we only have a test address on group 0, if we want to use the default address to deploy the script, do we need to export test addresses from other groups? Because the user might want to deploy the token on other groups, or we can deploy the contract on group 0 and then transfer token, but this need two transactions, WDYT?

Reply:
Yes, we should include a default address for each group, and only expose the group 0 one for external usage.

Reply:
> 1. We should use the default addresses to deploy the script and execute the script, so that the balance of recipient is not changed. Otherwise it's hard to test the balance changes

The balance still changes when minting tokens, we need to transfer the dust amount to the recipient here: https://github.com/alephium/alephium-web3/pull/258/commits/2d8542013c222e0a373080d07ca44e937b85fbab#diff-533abf520a4fdc1637791a205da159e5ce4c886c63fe8d00505e5961d47fd82aR84. Currently we do not support constructing transactions with inputs from different addresses.